src/System.Reflection.Metadata.ILDasm/src/Decoder/ILDecoder.cs

```diff
@@ -0,0 +1,582 @@
+        {
+            if (HasArgument(opCode))
+            {
+                if (_methodDefinition.Signature.Header.IsInstance)
+                {
+                    token--; //the first parameter is "this".
+                }
+                return _methodDefinition.GetParameter(token).Name;
+            }
+            return _methodDefinition.GetLocal(token).Name;
+            
+        }
+
+        private static string GetInlineTokenType(MetadataReader mdReader, int intOperand, ILTypeProvider provider)
+        {
+            if(IsMethodDefinition(intOperand) || IsMethodSpecification(intOperand) || IsMemberReference(intOperand))
+            {
+                return "method " + SolveMethodName(mdReader, intOperand, provider);
+            }
+            if (IsFieldDefinition(intOperand))
+            {
+                return "field " + GetFieldInformation(mdReader, intOperand, provider);
+            }
+            return GetTypeInformation(mdReader, intOperand, provider);
+        }
+
+        private static string GetTypeInformation(MetadataReader mdReader, int intOperand, ILTypeProvider provider)
+        {
+            if(IsTypeReference(intOperand))
+            {
+                var refHandle = MetadataTokens.TypeReferenceHandle(intOperand);
+                return SignatureDecoder.DecodeType(refHandle, provider).ToString();
+            }
+            if (IsTypeSpecification(intOperand))
+            {
+                var typeHandle = MetadataTokens.TypeSpecificationHandle(intOperand);
+                return SignatureDecoder.DecodeType(typeHandle, provider).ToString();
+            }
+            var defHandle = MetadataTokens.TypeDefinitionHandle(intOperand);
+            return SignatureDecoder.DecodeType(defHandle, provider).ToString();
+        }
+
+        private static ILInstruction CreateSwitchInstruction(ref BlobReader ilReader, int expectedSize, int ilOffset, OpCode opCode)
+        {
+            var caseNumber = ilReader.ReadUInt32();
+            int[] jumps = new int[caseNumber];
+            for(int i = 0; i < caseNumber; i++)
+            {
+                jumps[i] = ilReader.ReadInt32();
+            }
+            int size = 4 + expectedSize;
+            size += (int)caseNumber * 4;
+            return new ILSwitchInstruction(opCode, ilOffset, jumps, (int)caseNumber, caseNumber, size);
+        }
+
+        private static string GetArgumentString(MetadataReader mdReader, int intOperand, out bool isPrintable)
+        {
+            if (IsUserString(intOperand))
+            {
+                UserStringHandle usrStr = MetadataTokens.UserStringHandle(intOperand);
+                var str = mdReader.GetUserString(usrStr);
+                str = ProcessAndNromalizeString(str, out isPrintable);
+                return str;
+            }
+            throw new NotImplementedException("Argument String");
```

Comment:
This method should only be called with UserString token types, so I'm going to change that to InvalidArgumentException.